### DISCUSSION TITLE
Newb Question: Docker Install and Upgrade

### DISCUSSION THREAD

**Discussion by osseonews** [nocodb/nocodb#3605] [General]
I am a beginner with docker and had 2 questions. 

1. **Install:** In your instructions for docker you say: "to persist data in docker you can mount volume at /usr/app/data/ since 0.10.6." Your install instructions line: `-v "$(pwd)"/nocodb:/usr/app/data/ \` accomplishes that correct? So there is nothing else to add to install other than `-v "$(pwd)"/nocodb:/usr/app/data/ \` to persist data correctly?
2. **Upgrade:** On the upgrade instructions here: https://docs.nocodb.com/getting-started/upgrading/#docker - the -v is missing, so how does the app know where to pull the data from? Do we need to add the -v directive on the upgrade also, as the original container and image have been destroyed.

**Comment by wingkwong** [comment]
1. Mounting to your working directory is one of the ways to persist data. Apart from that, you can setup Storage plugins such as `AWS S3` to persist your data like attachment files. Other storage plugins can be found [here](https://docs.nocodb.com/setup-and-usages/app-store).
2. `-v` is not specific to upgrade. It should be based on your use case. Whenever you need to persist your data without using plugins, then you should specify it.

**Reply by osseonews** [reply]
Thanks. But, if we used -v when we first installed it, do we need to use it again on the upgrade? The original container is destroyed on the upgrade, so how would we access the persisted data without -v use on the upgrade?

**Reply by wingkwong** [reply]
Like I said before, whenever you need to persist your data without using plugins, then you should specify it. If you mount it in the first place, that means the files are available in your working directory. When you upgrade it, your container will be destroyed. Therefore, when you create a new one, you need to specify `-v` so that it can read the files back. So in your case, yes, you need to use it again.

**Reply by osseonews** [reply]
Ok, thanks. I got it. I guess I am just confused what exactly we would need to persist. Is this just for file attachments? I thought maybe there were core data features that needed persistence, but if we are not storing file attachments, what else is there to really persist?

**Reply by wingkwong** [reply]
For the data in the tables, that would be stored in your NC_DB. However, if you don't specify it, a local sqlite db will be created which is also under that directory. Files uploaded by users are also stored there.

**Reply by osseonews** [reply]
I'm confused. We are using postgres, so the data is presumably stored there. You mean that NC_DB replicates same data that is stored in the actual external database? I was under the impression that NC_DB is just an interface really on top of the database. Why would it store anything, assuming you are not using sqllite but some external database?

**Reply by wingkwong** [reply]
Actually there are different cases. What you said in the first part is also correct if you create your project by connecting to an external database. I think it would be clear if you look at the table in [here](https://docs.nocodb.com/engineering/architecture).

Here's the summary.

`NC_DB` is not specified -> local SQLite db.
Create Project without connecting to an external DB -> data will be stored in `NC_DB`. meta data will be stored in `NC_DB`
Create Project with connecting to an external DB -> data will be stored in an external DB. meta data will be stored in `NC_DB`
